### PR TITLE
ADBDEV-6613: Fix gpbackup/gprestore hanging on fatal error in goroutine

### PR DESCRIPTION
**Pull request by RekGRpth**
1) Make copy command cancelable

If the command used in copy does not support handling of SIGINT signal, then
such copy command cannot be canceled. timeout command does not handle SIGINT
signal by default, so its use in gprestore makes copy command cannot be
canceled until timeout command completes. This patch adds foreground argument
to timeout command so that it can handle SIGINT signal, and thus copy command
can be canceled while timeout command is running.

2) Fix gpbackup/gprestore hanging on fatal error in goroutine

gpbackup/gprestore would hang when a fatal error occurred in goroutine. This
was because fatal error in one goroutine was suppressed, while parallel
goroutines continued running. This patch adds cancel the context when recover
fatal error in goroutine, which causes the other goroutines to stop.

The test is added only for restore, and is not provided for backup, since it is
difficult to do without modifying the code.

---
This patch shouldn't be squashed!

### CODE REVIEW

**Review by Stolb27 — CHANGES_REQUESTED** [review]
Can you explain why we invent one more approach to sync go-routines instead of resuing existing one, provided in 55a7aa82c967f45562b36b268dfe26450f7c3986?

It's the same case as was described in the patch description:

> When using the jobs > 1 option, if a copy error occurred, the corresponding
goroutine was terminated. But other goroutines continued to wait for COPY and
restore froze.

**Comment by RekGRpth** [comment]
> Can you explain why we invent one more approach to sync go-routines instead of resuing existing one, provided in [55a7aa8](https://github.com/arenadata/gpbackup/commit/55a7aa82c967f45562b36b268dfe26450f7c3986)?
> 
> It's the same case as was described in the patch description:
> 
> > When using the jobs > 1 option, if a copy error occurred, the corresponding
> > goroutine was terminated. But other goroutines continued to wait for COPY and
> > restore froze.

I reworked solution

File: restore/data.go [hunk 1]
```diff
@@ -107,7 +106,10 @@ func restoreSingleTableData(queryContext context.Context, fpInfo *filepath.FileP
 		// was expected to set up
 		if backupConfig.SingleDataFile || resizeCluster {
 			agentErr := utils.CheckAgentErrorsOnSegments(globalCluster, globalFPInfo)
-			gplog.FatalOnError(agentErr)
+			if agentErr != nil {
+				gplog.Error(agentErr.Error())
+				return nil, agentErr
```
Comment (Stolb27) [inline restore/data.go:111]: It's the only kind of fatal error that may happens during table restore? I think the similar problem may happens otherwise. What do we do then? Will we add one more argument <some_kind_of>Err to return?

Comment (RekGRpth) [reply restore/data.go:111]: This is a special case only for helper errors, so that the goroutine can decide what to do next.

Comment (RekGRpth) [reply restore/data.go]: I simplified solution

Comment (Stolb27) [reply restore/data.go]: But you didn't answer on my question. gprestore still may hang in case of one of the go-routines fails with fatal error. Am I correct? Why not solve problem for any fatal error in go-routines?

Comment (RekGRpth) [reply restore/data.go]: > gprestore still may hang in case of one of the go-routines fails with fatal error. Am I correct?

It seems so.

File: restore/data.go [hunk 2]
```diff
@@ -301,7 +304,7 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
 					}
 				}
 				if err == nil {
-					err = restoreSingleTableData(ctx, &fpInfo, entry, tableName, whichConn)
+					err = restoreSingleTableData(ctx, cancel, &fpInfo, entry, tableName, whichConn)
```
Comment (Stolb27) [inline restore/data.go:307]: Passing context cancel function isn't recommended

Comment (RekGRpth) [reply restore/data.go]: Why? https://github.com/arenadata/gpbackup/blob/7087e491a1a01f0b71c49c31f48abd382f83dc10/restore/data.go#L264

File: restore/data.go (hunk 1 above)
Comment (RekGRpth) [reply restore/data.go]: I fixed it by adding cancel in recover

File: restore/data.go (hunk 2 above)
Comment (Stolb27) [reply restore/data.go]: PS I don't review patch made this change. So this part of question should be addressed to the author and reviewers.

I relied in my comment on the article from [here](https://p.agnihotry.com/post/understanding_the_context_package_in_golang/). As far I understand, passing cancel function as an argument violates responsibility principle, makes code more tightly couple and error prone.

Comment (RekGRpth) [reply restore/data.go]: > passing cancel function as an argument violates responsibility principle, makes code more tightly couple and error prone.

Then this is another argument against the current error file check every 5 seconds!

File: restore/data.go [hunk 3]
```diff
@@ -39,8 +39,7 @@ func CopyTableIn(queryContext context.Context, connectionPool *dbconn.DBConn, ta
 	if singleDataFile || resizeCluster {
 		//helper.go handles compression, so we don't want to set it here
 		customPipeThroughCommand = utils.DefaultPipeThroughProgram
-		errorFile := fmt.Sprintf("%s_error", globalFPInfo.GetSegmentPipePathForCopyCommand())
-		readFromDestinationCommand = fmt.Sprintf("(timeout 300 bash -c \"while [[ ! -p \"%s\" && ! -f \"%s\" ]]; do sleep 1; done\" || (echo \"Pipe not found %s\">&2; exit 1)) && %s", destinationToRead, errorFile, destinationToRead, readFromDestinationCommand)
+		readFromDestinationCommand = fmt.Sprintf("(timeout --foreground 300 bash -c \"while [ ! -p \"%s\" ]; do sleep 1; done\" || (echo \"Pipe not found %s\">&2; exit 1)) && %s", destinationToRead, destinationToRead, readFromDestinationCommand)
```
Comment (Stolb27) [inline restore/data.go:42]: Why should we not check error file during waiting on COPY anymore? There is lack of details in the PR description. How this changes connected with remaining patch?

Also, I strongly recommend to separate adding foreground option to the separate commit with detailed description about timeout behavior.

Comment (RekGRpth) [reply restore/data.go:42]: > I strongly recommend to separate adding foreground option to the separate commit with detailed description about timeout behavior.

Unfortunately, neither adding context cancellation to the recover nor adding the foreground argument separately solves the hanging problem.

Comment (RekGRpth) [reply restore/data.go]: > Why should we not check error file during waiting on COPY anymore? There is lack of details in the PR description. How this changes connected with remaining patch?

I restored checking error files during waiting timeout

Comment (RekGRpth) [reply restore/data.go]: > > I strongly recommend to separate adding foreground option to the separate commit with detailed description about timeout behavior.
> 
> Unfortunately, neither adding context cancellation to the recover nor adding the foreground argument separately solves the hanging problem.

Maybe split the patch into two commits? The first one will add the `foreground` argument to the `timeout` command, and the second one will add context `cancel` on `recover` and add a new test for it?

Comment (Stolb27) [reply restore/data.go]: > Maybe split the patch into two commits?

> I strongly recommend to separate adding foreground option to the separate **commit**

Exactly.

File: restore/data.go (hunk 2 above)
Comment (Stolb27) [reply restore/data.go]: But it's a problem for the separate task. Discussed one was fixed.
